src/L2Penalty_alg.jl

```diff
@@ -0,0 +1,431 @@
+		set_solver_specific!(stats,:smooth_obj,fx)
+		set_solver_specific!(stats,:nonsmooth_obj, hx)
+		set_solver_specific!(stats, :theta, sqrt_θ)
+
+		set_status!(
+      stats,
+      get_status(
+        nlp,
+        elapsed_time = stats.elapsed_time,
+				n_iter_since_decrease = n_iter_since_decrease,
+        iter = stats.iter,
+        optimal = solved,
+        max_eval = max_eval,
+        max_time = max_time,
+        max_iter = max_iter,
+				max_decreas_iter = max_decreas_iter
+      ),
+    )
+
+		callback(nlp, solver, stats)
+
+		done = stats.status != :unknown
+	end
+
+	set_solution!(stats, x)
+	return stats
+
+end
+
+function get_status(
+  nlp::M;
+  elapsed_time = 0.0,
+  iter = 0,
+  optimal = false,
+	n_iter_since_decrease = 0,
+  max_eval = Inf,
+  max_time = Inf,
+  max_iter = Inf,
+	max_decreas_iter = Inf
+) where{ M <: AbstractNLPModel }
+  if optimal
+    :first_order
+  elseif iter > max_iter
+    :max_iter
+  elseif elapsed_time > max_time
+    :max_time
+  elseif neval_obj(nlp) > max_eval && max_eval > -1
+    :max_eval
+	elseif n_iter_since_decrease ≥ max_decreas_iter
+		:infeasible
+	else
+    :unknown
+  end
+end
+
+mutable struct L2_R2N_subsolver{T <: Real, V <: AbstractVector{T}} <: AbstractOptimizationSolver
+  u1::V
+	u2::V
+end
+
+function L2_R2N_subsolver(
+  reg_nlp::AbstractRegularizedNLPModel{T, V};
+	) where{T, V}
+	x0 = reg_nlp.model.meta.x0
+	n = reg_nlp.model.meta.nvar
+	m = length(reg_nlp.h.b)
+	#x = zero(x0)
+	u1 = similar(x0, n+m)
+	u2 = zeros(eltype(x0), n+m)
+
+
+  return L2_R2N_subsolver(
+		u1,
+		u2,
+	)
+end
+
+function solve!(
+	solver::L2_R2N_subsolver{T, V},
+	reg_nlp::AbstractRegularizedNLPModel{T, V},
+	stats::GenericExecutionStats{T, V, V, Any},
+	∇fk::V,
+	Q::L,
+	σk::T;
+	x = reg_nlp.model.meta.x0,
+	atol = eps(T)^(0.5),
+	max_time = T(30),
+	max_iter = 10000
+	) where{T <: Real, V <: AbstractVector{T} , L <: AbstractLinearOperator}
+
+	start_time = time()
+	set_time!(stats, 0.0)
+	set_iter!(stats, 0)
+	
+	n = reg_nlp.model.meta.nvar
+	m = length(reg_nlp.h.b)
+	Δ = reg_nlp.h.h.lambda/σk
+
+	u1 = solver.u1
+	u2 = solver.u2
+
+	# Create problem
+	@. u1[1:n] = ∇fk/σk	
+	@. u1[n+1:n+m] = -reg_nlp.h.b
+
+	αₖ = 0.0
+
+	H1 = [-Q reg_nlp.h.A']
+	H2 = [reg_nlp.h.A αₖ*opEye(m,m)]
+	H = [H1;H2]
+	x1,_ = minres_qlp(H,u1)
+	#println(stats_minres)
+	if norm(x1[n+1:n+m]) <= Δ
+		set_solution!(stats,x1[1:n])
+		return
+	end
+	u2[n+1:n+m] .= x1[n+1:n+m]
+	x2,_ = minres_qlp(H,u2)
+	αₖ += norm(x1[n+1:n+m])^2/(x1[n+1:n+m]'x2[n+1:n+m])*(norm(x1[n+1:n+m])- Δ)/Δ
+	k = 0
+
+	while abs(norm(x1[n+1:n+m]) - Δ) > eps(T)^(0.75) && stats.iter < max_iter && stats.elapsed_time < max_time
+		H2 = [reg_nlp.h.A αₖ*opEye(m,m)]
+		H = [H1;H2]
+
+		x1,_ = minres_qlp(H,u1)
+		u2[n+1:n+m] .= x1[n+1:n+m]
+		x2,_ = minres_qlp(H,u2)
+		αₖ += norm(x1[n+1:n+m])^2/(x1[n+1:n+m]'x2[n+1:n+m])*(norm(x1[n+1:n+m])- Δ)/Δ
+		set_iter!(stats,stats.iter + 1)
+		set_time!(stats,time()-start_time)
+	end
+	set_solution!(stats,x1[1:n])
+
+end
```

Comment:
Shouldn't this function be in ShiftedProximalOperators?

Reply:
I don't really know. 

I think it should be considered as a JSO solver since this is what R2N will expect as a subsolver.
Therefore, it should be written here instead of ShiftedProximalOperators but I am open to rewrite it elsewhere if necessary.